src/encoded/upgrade/genetic_modification.py

```diff
@@ -185,4 +185,78 @@ def genetic_modification_5_6(value, system):
 def genetic_modification_6_7(value, system):
     # https://encodedcc.atlassian.net/browse/ENCD-4042
     if value['purpose'] == 'validation':
-        value['purpose'] = 'characterization'
\ No newline at end of file
+        value['purpose'] = 'characterization'
+
+
+@upgrade_step('genetic_modification', '7', '8')
+def genetic_modification_7_8(value, system):
+    # https://encodedcc.atlassian.net/browse/ENCD-4349
+    import re
+    import copy
+    import time
+    import json
+
+    conn = system['registry'][CONNECTION]
+
+    sources = {
+        'addgene': r'^\d{5,6}$',
+        'bacpac': r'^([A-Z]{2,3}\d{2,3}|[A-Z]{3})-\d{1,4}[A-Z]\d{1,2}$',
+        'dharmacon': r'^[DL]-\d{6}-\d{2}(-\d{2,4})?$',
+        'hugo-bellen': r'^MI\d{5}$',
+        'human-orfeome': r'^([A-Z]{2})?\d{1,9}$',
+        'plasmid-repository': r'^HsCD\d{8}$',
+        'sigma': r'^[A-Z]{3}\d{3}$',
+        'source-bioscience': r'^[A-Z]{3}\d{3,4}[a-z][A-Z]\d{2}(_[A-Z]\d{2})?$',
+        'thermo-fisher': r'^[a-zA-Z]{1,3}\d{5,6}$',
+        'trc': r'^TRCN\d{10}$',
+        'brenton-graveley': r'^BGC#\d{7}$'
+    }
+    reagents = value.get('reagents')
+    if reagents:
+        new_reagents = []
+        unmatched_reagents = ''
+        for reagent in reagents:
+            identifier = reagent['identifier']
+            source_id = reagent['source']
+            source_obj = conn.get_by_uuid(source_id)
+            source_name = source_obj.properties['name']
+            new_reagent = copy.deepcopy(reagent)
+            matching = []
+            source_from_reagent = None
+            for source in sources:
+                if source in source_name:
+                    source_from_reagent = source_name
+                    break
+            # Preferentially add prefix to reagent that matches the reagent source
+            if source_from_reagent is not None:
+                regex = sources[source_from_reagent]
```

Comment:
I think this block could be added to the above for directly.  Then if the for loop exits without setting source_from_reagent we can 
```
if not source_from_reagent:
    for source, regex in sources.items():
        ....2nd for loop check
```
Also, we don't need to break in the second for loop like we do in the first?  The second could add multiple values.

And do we need to specifically check is not None in all these cases?

'regex' is not a good variable name, pretty generic.  'src_regex' would be better.

Reply:
1. & 2. ) The loops definitely could be consolidated. The second loop shouldn't break because it is possible for the identifier to match multiple and we want to catch all of them.
3. ) re.match returns None if there is no match, so checking 'if re.match(...)' should suffice. I'll change this
4. ) I'll change regex to src_regex.